app/controllers/calendar_controller.rb

```diff
@@ -1,5 +1,27 @@
+require 'google/apis/calendar_v3'
+require 'google/api_client/client_secrets'
+
 class CalendarController < ApplicationController
 
+  before_action :check_authentication, except: :authenticate
+  before_action :refresh_token
+
+  def authenticate
+    if params['code'] == nil
+      redirect_to GoogleOauth.request_code_uri
```

Comment:
would it make sense to do:
`fail ExceptionOfYourChoice, 'Exception message' if params['code'].nil?`

and handle this in the rescue section ?